server/container/util-java8/src/main/java/org/apache/james/util/CompletableFutureUtil.java

```diff
@@ -51,17 +58,29 @@
         return futurStream
             .thenCompose(stream ->
                 CompletableFutureUtil.allOf(
-                    stream.map(action::apply)));
+                    stream.map(action)));
     }
 
     public static <T, U> CompletableFuture<Stream<U>> map(CompletableFuture<Stream<T>> futurStream, Function<T, U> action) {
         return futurStream
             .thenApply(stream ->
-                stream.map(value ->
-                    action.apply(value)));
+                stream.map(action));
+    }
+
+    public static <T, U> CompletableFuture<Optional<T>> reduce(BinaryOperator<T> binaryOperator, CompletableFuture<Stream<T>> futureStream) {
+        return futureStream.thenApply(stream -> stream.reduce(binaryOperator));
     }
 
     public static <T> CompletableFuture<T> keepValue(Supplier<CompletableFuture<Void>> supplier, T value) {
         return supplier.get().thenApply(any -> value);
     }
+
+    public static <T> Function<Boolean, CompletableFuture<Boolean>> composeIfTrue(Supplier<CompletableFuture<T>> composeOperation) {
```

Comment:
I don't have the solution at hand, keep it that way if you want until we learn our FP